xymond/hobbitd_filestore.8
.TH HOBBITD_FILESTORE 8 "Version 4.2.1: 27 Nov 2008" "Xymon"
.SH NAME
hobbitd_filestore \- hobbitd worker module for storing Xymon data
.SH SYNOPSIS
.B "hobbitd_channel --channel=status hobbitd_filestore --status [options]"
.br
.B "hobbitd_channel --channel=data   hobbitd_filestore --data [options]"
.br
.B "hobbitd_channel --channel=notes  hobbitd_filestore --notes [options]"
.br
.B "hobbitd_channel --channel=enadis hobbitd_filestore --enadis [options]"

.SH DESCRIPTION
hobbitd_filestore is a worker module for hobbitd, and as such it is normally
run via the
.I hobbitd_channel(8)
program. It receives hobbitd messages from a hobbitd channel via stdin, and 
stores these in the filesystem in a manner that is compatible with the 
Big Brother daemon, bbd.

This program can be started multiple times, if you want to store
messages for more than one channel.

.SH OPTIONS
.IP "--status"
Incoming messages are "status" messages, they will be stored in the
$BBLOGS/ directory. If you are using 
.I xymon(7)
throughout your Xymon server, you will not need to run this
module to save status messages, unless you have a third-party 
add-on that reads the status-logs directly.
This module is NOT needed to get trend graphs, you should run the 
.I hobbitd_rrd(8)
module instead.

.IP "--data"
Incoming messages are "data" messages, they will be stored in the
$BBDATA directory. This module is not needed, unless you have a
third-party module that processes the data-files. This module is
NOT needed to get trend graphs, you should run the 
.I hobbitd_rrd(8)
module instead.

.IP "--notes"
Incoming messages are "notes" messages, they will be stored in the
$BBNOTES directory. This modules is only needed if you want to 
allow people to remotely update the notes-files available on the
Xymon webpages.

.IP "--enadis"
Incoming messages are enable/disable messages, they will update 
files in the $BBDISABLED directory. This is only needed if you have
third-party add-ons that use these files.

.IP "--dir=DIRECTORY"
Overrides the default output directory.

.IP "--html"
Used together with "--status". Tells hobbitd_filestore to also save
an HTML version of the status-log. Should not be used unless you 
must run with "BBLOGSTATUS=static".

.IP "--htmldir=DIRECTORY"
The directory where HTML-versions of the status logs are stored.
Default: $BBHTML

.IP "--htmlext=.EXT"
Set the filename extension for generated HTML files. By default, HTML
files are saved with a ".html" extension.

.IP "--multigraphs=TEST1[,TEST2]"
This causes hobbitd_filestore to generate HTML status pages with links to 
service graphs that are split up into multiple images, with at most 5 graphs
per image. If not specified, only the "disk" status is split up this way.

.IP "--only=test[,test,test]"
Save status messages only for the listed set of tests. This can be useful
if you have an external script that needs to parse some of the status logs,
but you do not want to save all status logs.

.IP "--debug"
Enable debugging output.

.SH FILES
This module does not rely on any configuration files.

.SH "SEE ALSO"
hobbitd_channel(8), hobbitd_rrd(8), hobbitd(8), xymon(7)



Changes
Changes from 4.3.0 RC 1 -> 4.3.0 (4 Mar 2011)
=============================================
* Critical view and other webpages: Make the 'All systems OK' message 
  configurable. Also allow the header/footer for the Critical Systems 
  view to be configurable.
  Suggestion and preliminary patch from Buchan Milne.
* xymonnet: Improve error report when HTTP tests get an empty response - 
  'HTTP error 0' sounds weird.
* report / snapshot CGI's: Fix buffer overrun in the HTML delimiter 
  generated in the "please wait..." message. Also, fix potential buffer 
  overrun in report CGI if invoked with a large value for the "style" 
  parameter. 
  Reported by Rolf Biesbroek.
* Graph definitions (graphs.cfg): Multi graphs cannot use a regex pattern. 
  Problem report by Brian Majeska
* Solaris interface statistics: Filter out "mac" and "wrsmd" devices at
  the client side. 
  Update RRD handler to also filter "wrsmd" at the server side, like we 
  already did for "mac" devices. 
  Cf. http://www.xymon.com/archive/2009/06/msg00204.html
* Documentation: Document the XMH_* fields available in xymondboard commands.
* Documentation: Document SPLITNCV and "trends" methods of doing 
  custom graphs.
* RRD definitions: Allow override of --step/-s option for rrdcreate, 
  from template supplied in rrddefinitions.cfg. 
  Suggestion from Brian Majeska.
* mailack: Remove restriction on how long a subjectline/message body can be.
* Build procedure: Add notice about running upgrade script before 
  installing the new version.
* xymond_alert: Document --trace option
* Alerts: For recovery messages, add information so you can tell 
  whether the recovery was due to the service actually recovering, or 
  if it was merely disabled.
* xymond_alert: Fix missing element in array of alert status texts used 
  for tracing. 
  Spotted by Dominique Frise.
* Add support for FreeBSD v8 modified ifstat output
* Documentation: Update information about the Xymon mailing lists 
  following move to Mailman and new archive URL.
* HP/UX client: Use "swapinfo" to extract memory utilisation data, 
  instead of the hpux-meminfo utility. 
  By Earl Flack http://lists.xymon.com/pipermail/xymon/2010-December/030100.html

Changes from 4.3.0 beta 3 -> 4.3.0 RC 1 (23 Jan 2011)
=====================================================
* hosts.cfg badldap documentation: Document that for LDAP URL's you must 
  use 'badldapurl'. Reported by Simo Hmami.
* xymond flap detection: Make number of tracked status changes and the 
  flap-check period configurable. Change the defaults to trigger flapping 
  at more than 5 status changes in a 30 minute period.
* sendmessage: Enhanced error reporting, to help track down communication 
  problems.
* xymond_client: Fix Windows SVC status handling to avoid coredumps, 
  memory corruption and other nasties. 
  Will now report the real name of the service, instead of the pattern used in 
  the analysis.cfg file.
  NOTE: Slight change to status message format.
* Client handler: Fix owner/user check parsing. Reported by Ian Marsh 
  http://www.xymon.com/archive/2011/01/msg00133.html (also broken in 4.2.3).
* xymongen: Fix broken --doc-window option handling. Reported by Tom Schmitt.
* Xymongen: Fix documentation of the --doc-window/--no-doc-window options.
* Webpage background: Use a CSS and a new set of gif's to implement a 
  background that works on all displays, regardless of width. Uses a new 
  xymonbody.css stylesheet which can also control some other aspects of the 
  webpage. From Francois Claire.
* Documentation: The xymon 'rename' command should be used AFTER renaming 
  a host in hosts.cfg, not before. From Tom Georgoulias.
* Memory status: Add some sanity checks for the memory utilisation reported 
  by clients. Occasionally we get completely bogus data from clients, so only 
  act on them if percentages do not exceed 100.
* Critical systems view: Add "--tooltips" option so you can save screen space 
  by hiding the host descriptions in a tooltip, like we do on the statically 
  generated pages. Feature request from Chris Morris.
* Solaris client: Report "swap -l" in addition to "swap -s" for swap usage. 
  Backend prefers output from "swap -l" when determining swap utilisation.
* Webpage menu: Use the CSS and GIF's by Malcolm Hunter - they are much nicer 
  than the ones from Debian. Distribute both the blue and the grey version, 
  and configure which one to use in xymonserver.cfg.
* Graph zoom: Use float variables when calculating the upper/lower limits 
  of the graph. Fixes vertical zoom.
* xymond: Make sure we do not perform socket operations on invalid sockets 
  (e.g. those from a scheduled task pseudo-connection)
* Installation: Remove any existing old commands before creating symlinks
* xymonproxy: Fix broken compatibility option '--bbdisplay'
* Fix eventlog summary/count enums so they dont clash with Solaris predefined 
  entities
* History- and hostdata-modules: Dont save data if there is less than 5% 
  free space on the filesystem. Also, dont save hostdata info more than 5 
  times per hour.
* Historical statuslog display: Work-around for crash when status-log is 
  empty
* fping.sh configure sub-script: Fix syntax error in suggested 'sudoers' 
  configuration, and point to the found fping binary. From Steff Coene.
* namematch routine: Fix broken matching when doing simple matching against 
  two strings where one was a subset of the other. 
  http://www.xymon.com/archive/2010/11/msg00177.html .  Reported by Elmar Heeb 
  who also provided a patch, although I chose a different solution to this.
* Xymon net: Fix broken compile when LDAP-checks are disabled. Reported by 
  Roland Soderstrom, fix from Ralph Mitchell.
* xymon(7) manpage: Drop notice that renaming in 4.3.0 is not complete
* Installation: Setup links for the commonly used Hobbit binaries 
  (bb, bbcmd, bbdigest, bbhostgrep, bbhostshow)
* Upgrade script: Setup symlinks for the old names of the standard webpages

hobbitd/hobbitd_filestore.c
/*----------------------------------------------------------------------------*/
/* Xymon message daemon.                                                      */
/*                                                                            */
/* This is a hobbitd worker module, it should be run off hobbitd_channel.     */
/*                                                                            */
/* This module implements the traditional Big Brother filebased storage of    */
/* incoming status messages to the bbvar/logs/, bbvar/data/, bb/www/notes/    */
/* and bbvar/disabled/ directories.                                           */
/*                                                                            */
/* Copyright (C) 2004-2009 Henrik Storner <henrik@hswn.dk>                    */
/*                                                                            */
/* This program is released under the GNU General Public License (GPL),       */
/* version 2. See the file "COPYING" for details.                             */
/*                                                                            */
/*----------------------------------------------------------------------------*/

static char rcsid[] = "$Id$";

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <utime.h>
#include <dirent.h>
#include <limits.h>
#include <signal.h>
#include <errno.h>
#include <libgen.h>

#include "libbbgen.h"

#include "hobbitd_worker.h"

static char *multigraphs = ",disk,inode,qtree,quotas,snapshot,TblSpace,if_load,";
static int locatorbased = 0;

enum role_t { ROLE_STATUS, ROLE_DATA, ROLE_NOTES, ROLE_ENADIS};

void update_file(char *fn, char *mode, char *msg, time_t expire, char *sender, time_t timesincechange, int seq)
{
	FILE *logfd;
	char tmpfn[PATH_MAX];
	char *p;

	MEMDEFINE(tmpfn);

	dbgprintf("Updating seq %d file %s\n", seq, fn);

	p = strrchr(fn, '/');
	if (p) {
		*p = '\0';
		sprintf(tmpfn, "%s/.%s", fn, p+1);
		*p = '/';
	}
	else {
		sprintf(tmpfn, ".%s", fn);
	}

	logfd = fopen(tmpfn, mode);
	fwrite(msg, strlen(msg), 1, logfd);
	if (sender) fprintf(logfd, "\n\nMessage received from %s\n", sender);
	if (timesincechange >= 0) {
		char timestr[100];
		char *p = timestr;
		if (timesincechange > 86400) p += sprintf(p, "%ld days, ", (timesincechange / 86400));
		p += sprintf(p, "%ld hours, %ld minutes", 
				((timesincechange % 86400) / 3600), ((timesincechange % 3600) / 60));
		fprintf(logfd, "Status unchanged in %s\n", timestr);
	}
	fclose(logfd);

	if (expire) {
		struct utimbuf logtime;
		logtime.actime = logtime.modtime = expire;
		utime(tmpfn, &logtime);
	}

	rename(tmpfn, fn);

	MEMUNDEFINE(tmpfn);
}

void update_htmlfile(char *fn, char *msg, 
		     char *hostname, char *service, int color, int flapping,
		     char *sender, char *flags,
		     time_t logtime, time_t timesincechange,
		     time_t acktime, char *ackmsg,
                     time_t disabletime, char *dismsg)
{
	FILE *output;
	char *tmpfn;
	char *firstline, *restofmsg;
	char *displayname = hostname;
	char *ip = "";
	char timestr[100];

	MEMDEFINE(timestr);

	tmpfn = (char *) malloc(strlen(fn)+5);
	sprintf(tmpfn, "%s.tmp", fn);
	output = fopen(tmpfn, "w");

	if (output) {
		firstline = msg;
		restofmsg = strchr(msg, '\n');
		if (restofmsg) {
			*restofmsg = '\0';
			restofmsg++;
		}
		else {
			restofmsg = "";
		}

		if (timesincechange >= 0) {
			char *p = timestr;
			if (timesincechange > 86400) p += sprintf(p, "%ld days, ", (timesincechange / 86400));
			p += sprintf(p, "%ld hours, %ld minutes", 
					((timesincechange % 86400) / 3600), ((timesincechange % 3600) / 60));
		}

		generate_html_log(hostname, displayname, service, ip,
			color, flapping, sender, flags,
			logtime, timestr,
			firstline, restofmsg, NULL,
			acktime, ackmsg, NULL,
			disabletime, dismsg,
			0, 1, 0, locatorbased, multigraphs, NULL, 
			NULL, NULL, NULL,
			0,
			output);

		fclose(output);
		rename(tmpfn, fn);
	}

	xfree(tmpfn);
	MEMUNDEFINE(timestr);
}

void update_enable(char *fn, time_t expiretime)
{
	time_t now = getcurrenttime(NULL);

	dbgprintf("Enable/disable file %s, time %d\n", fn, (int)expiretime);

	if (expiretime <= now) {
		if (unlink(fn) != 0) {
			errprintf("Could not remove disable-file '%s':%s\n", fn, strerror(errno));
		}
	}
	else {
		FILE *enablefd;
		struct utimbuf logtime;

		enablefd = fopen(fn, "w");
		if (enablefd) {
			fclose(enablefd);
		}

		logtime.actime = logtime.modtime = expiretime;
		utime(fn, &logtime);
	}
}

static int wantedtest(char *wanted, char *key)
{
	char *p, *ckey;

	if (wanted == NULL) return 1;

	ckey = (char *)malloc(strlen(key) + 3);
	sprintf(ckey, ",%s,", key);
	p = strstr(wanted, ckey);
	xfree(ckey);

	if (p) dbgprintf("wantedtest: Found '%s' at '%s'\n", key, p);
	else dbgprintf("wantedtest: '%s' not found\n", key);

	return (p != NULL);
}

int main(int argc, char *argv[])
{
	char *filedir = NULL;
	char *htmldir = NULL;
	char *htmlextension = "html";
	char *onlytests = NULL;
	char *msg;
	enum role_t role = ROLE_STATUS;
	enum msgchannels_t chnid = C_STATUS;
	int argi;
	int seq;
	int running = 1;

	/* Dont save the error buffer */
	save_errbuf = 0;

	for (argi = 1; (argi < argc); argi++) {
		if (strcmp(argv[argi], "--status") == 0) {
			role = ROLE_STATUS;
			chnid = C_STATUS;
			if (!filedir) filedir = xgetenv("BBLOGS");
		}
		else if (strcmp(argv[argi], "--html") == 0) {
			role = ROLE_STATUS;
			chnid = C_STATUS;
			if (!htmldir) htmldir = xgetenv("BBHTML");
		}
		else if (strcmp(argv[argi], "--data") == 0) {
			role = ROLE_DATA;
			chnid = C_DATA;
			if (!filedir) filedir = xgetenv("BBDATA");
		}
		else if (strcmp(argv[argi], "--notes") == 0) {
			role = ROLE_NOTES;
			chnid = C_NOTES;
			if (!filedir) filedir = xgetenv("BBNOTES");
		}
		else if (strcmp(argv[argi], "--enadis") == 0) {
			role = ROLE_ENADIS;
			chnid = C_ENADIS;
			if (!filedir) filedir = xgetenv("BBDISABLED");
		}
		else if (strcmp(argv[argi], "--debug") == 0) {
			debug = 1;
		}
		else if (argnmatch(argv[argi], "--dir=")) {
			filedir = strchr(argv[argi], '=')+1;
		}
		else if (argnmatch(argv[argi], "--htmldir=")) {
			htmldir = strchr(argv[argi], '=')+1;
		}
		else if (argnmatch(argv[argi], "--htmlext=")) {
			htmlextension = strchr(argv[argi], '=')+1;
		}
		else if (argnmatch(argv[argi], "--only=")) {
			char *p = strchr(argv[argi], '=') + 1;
			onlytests = (char *)malloc(3 + strlen(p));
			sprintf(onlytests, ",%s,", p);
		}
		else if (argnmatch(argv[argi], "--multigraphs=")) {
			char *p = strchr(argv[argi], '=');
			multigraphs = (char *)malloc(strlen(p+1) + 3);
			sprintf(multigraphs, ",%s,", p+1);
		}
		else if (argnmatch(argv[argi], "--locator=")) {
			char *p = strchr(argv[argi], '=');
			locator_init(p+1);
			locatorbased = 1;
		}
	}

	if (filedir == NULL) {
		errprintf("No directory given, aborting\n");
		return 1;
	}

	/* For picking up lost children */
	setup_signalhandler("hobbitd_filestore");
	signal(SIGPIPE, SIG_DFL);

	if (onlytests) dbgprintf("Storing tests '%s' only\n", onlytests);
	else dbgprintf("Storing all tests\n");

	while (running) {
		char *metadata[20] = { NULL, };
		char *statusdata = "";
		char *p;
		int metacount;
		char *hostname, *testname;
		time_t expiretime = 0;
		char logfn[PATH_MAX];

		MEMDEFINE(logfn);

		msg = get_hobbitd_message(chnid, "filestore", &seq, NULL);
		if (msg == NULL) {
			running = 0;
			MEMUNDEFINE(logfn);
			continue;
		}

		p = strchr(msg, '\n'); 
		if (p) {
			*p = '\0'; 
			statusdata = p+1;
		}

		metacount = 0;
		memset(&metadata, 0, sizeof(metadata));
		p = gettok(msg, "|");
		while (p && (metacount < 20)) {
			metadata[metacount++] = p;
			p = gettok(NULL, "|");
		}

		if ((role == ROLE_STATUS) && (metacount >= 14) && (strncmp(metadata[0], "@@status", 8) == 0)) {
			/* @@status|timestamp|sender|origin|hostname|testname|expiretime|color|testflags|prevcolor|changetime|ackexpiretime|ackmessage|disableexpiretime|disablemessage|clientmsgtstamp|flapping */
			int ltime, flapping = 0;
			time_t logtime = 0, timesincechange = 0, acktime = 0, disabletime = 0;

			hostname = metadata[4];
			testname = metadata[5];
			if (!wantedtest(onlytests, testname)) {
				dbgprintf("Status dropped - not wanted\n");
				MEMUNDEFINE(logfn);
				continue;
			}

			sprintf(logfn, "%s/%s.%s", filedir, commafy(hostname), testname);
			expiretime = atoi(metadata[6]);
			statusdata = msg_data(statusdata);
			sscanf(metadata[1], "%d.%*d", &ltime); logtime = ltime;
			timesincechange = logtime - atoi(metadata[10]);
			update_file(logfn, "w", statusdata, expiretime, metadata[2], timesincechange, seq);
			if (htmldir) {
				char *ackmsg = NULL;
				char *dismsg = NULL;
				char htmllogfn[PATH_MAX];

				MEMDEFINE(htmllogfn);

				if (metadata[11]) acktime = atoi(metadata[11]);
				if (metadata[12] && strlen(metadata[12])) ackmsg = metadata[12];
				if (ackmsg) nldecode(ackmsg);

				if (metadata[13]) disabletime = atoi(metadata[13]);
				if (metadata[14] && strlen(metadata[14]) && (disabletime > 0)) dismsg = metadata[14];
				if (dismsg) nldecode(dismsg);

				flapping = (metadata[16] ? (*metadata[16] == '1') : 0);

				sprintf(htmllogfn, "%s/%s.%s.%s", htmldir, hostname, testname, htmlextension);
				update_htmlfile(htmllogfn, statusdata, hostname, testname, parse_color(metadata[7]), flapping,
						     metadata[2], metadata[8], logtime, timesincechange, 
						     acktime, ackmsg,
						     disabletime, dismsg);

				MEMUNDEFINE(htmllogfn);
			}
		}
		else if ((role == ROLE_DATA) && (metacount > 5) && (strncmp(metadata[0], "@@data", 6) == 0)) {
			/* @@data|timestamp|sender|hostname|testname */
			p = hostname = metadata[4]; while ((p = strchr(p, '.')) != NULL) *p = ',';
			testname = metadata[5];
			if (!wantedtest(onlytests, testname)) {
				dbgprintf("data dropped - not wanted\n");
				MEMUNDEFINE(logfn);
				continue;
			}

			statusdata = msg_data(statusdata); if (*statusdata == '\n') statusdata++;
			sprintf(logfn, "%s/%s.%s", filedir, hostname, testname);
			expiretime = 0;
			update_file(logfn, "a", statusdata, expiretime, NULL, -1, seq);
		}
		else if ((role == ROLE_NOTES) && (metacount > 3) && (strncmp(metadata[0], "@@notes", 7) == 0)) {
			/* @@notes|timestamp|sender|hostname */
			hostname = metadata[3];
			statusdata = msg_data(statusdata); if (*statusdata == '\n') statusdata++;
			sprintf(logfn, "%s/%s", basename(filedir), hostname);
			expiretime = 0;
			update_file(logfn, "w", statusdata, expiretime, NULL, -1, seq);
		}
		else if ((role == ROLE_ENADIS) && (metacount > 5) && (strncmp(metadata[0], "@@enadis", 8) == 0)) {
			/* @@enadis|timestamp|sender|hostname|testname|expiretime */
			p = hostname = metadata[3]; while ((p = strchr(p, '.')) != NULL) *p = ',';
			testname = metadata[4];
			expiretime = atoi(metadata[5]);
			sprintf(logfn, "%s/%s.%s", filedir, hostname, testname);
			update_enable(logfn, expiretime);
		}
		else if (((role == ROLE_STATUS) || (role == ROLE_DATA) || (role == ROLE_ENADIS)) && (metacount > 3) && (strncmp(metadata[0], "@@drophost", 10) == 0)) {
			/* @@drophost|timestamp|sender|hostname */
			DIR *dirfd;
			struct dirent *de;
			char *hostlead;

			p = hostname = metadata[3]; while ((p = strchr(p, '.')) != NULL) *p = ',';
			hostlead = malloc(strlen(hostname) + 2);
			strcpy(hostlead, hostname); strcat(hostlead, ".");

			dirfd = opendir(filedir);
			if (dirfd) {
				while ( (de = readdir(dirfd)) != NULL) {
					if (strncmp(de->d_name, hostlead, strlen(hostlead)) == 0) {
						sprintf(logfn, "%s/%s", filedir, de->d_name);
						unlink(logfn);
					}
				}
				closedir(dirfd);
			}

			xfree(hostlead);
		}
		else if (((role == ROLE_STATUS) || (role == ROLE_DATA) || (role == ROLE_ENADIS)) && (metacount > 4) && (strncmp(metadata[0], "@@droptest", 10) == 0)) {
			/* @@droptest|timestamp|sender|hostname|testname */
			p = hostname = metadata[3]; while ((p = strchr(p, '.')) != NULL) *p = ',';
			testname = metadata[4];
			sprintf(logfn, "%s/%s.%s", filedir, hostname, testname);
			unlink(logfn);
		}
		else if (((role == ROLE_STATUS) || (role == ROLE_DATA) || (role == ROLE_ENADIS)) && (metacount > 4) && (strncmp(metadata[0], "@@renamehost", 12) == 0)) {
			/* @@renamehost|timestamp|sender|hostname|newhostname */
			DIR *dirfd;
			struct dirent *de;
			char *hostlead;
			char *newhostname;
			char newlogfn[PATH_MAX];

			MEMDEFINE(newlogfn);

			p = hostname = metadata[3]; while ((p = strchr(p, '.')) != NULL) *p = ',';
			hostlead = malloc(strlen(hostname) + 2);
			strcpy(hostlead, hostname); strcat(hostlead, ".");
			p = newhostname = metadata[4]; while ((p = strchr(p, '.')) != NULL) *p = ',';

			dirfd = opendir(filedir);
			if (dirfd) {
				while ( (de = readdir(dirfd)) != NULL) {
					if (strncmp(de->d_name, hostlead, strlen(hostlead)) == 0) {
						char *testname = strchr(de->d_name, '.');
						sprintf(logfn, "%s/%s", filedir, de->d_name);
						sprintf(newlogfn, "%s/%s%s", filedir, newhostname, testname);
						rename(logfn, newlogfn);
					}
				}
				closedir(dirfd);
			}
			xfree(hostlead);

			MEMUNDEFINE(newlogfn);
		}
		else if (((role == ROLE_STATUS) || (role == ROLE_DATA) || (role == ROLE_ENADIS)) && (metacount > 5) && (strncmp(metadata[0], "@@renametest", 12) == 0)) {
			/* @@renametest|timestamp|sender|hostname|oldtestname|newtestname */
			char *newtestname;
			char newfn[PATH_MAX];

			MEMDEFINE(newfn);

			p = hostname = metadata[3]; while ((p = strchr(p, '.')) != NULL) *p = ',';
			testname = metadata[4];
			newtestname = metadata[5];
			sprintf(logfn, "%s/%s.%s", filedir, hostname, testname);
			sprintf(newfn, "%s/%s.%s", filedir, hostname, newtestname);
			rename(logfn, newfn);

			MEMUNDEFINE(newfn);
		}
		else if (strncmp(metadata[0], "@@shutdown", 10) == 0) {
			running = 0;
		}
		else if (strncmp(metadata[0], "@@logrotate", 11) == 0) {
			char *fn = xgetenv("HOBBITCHANNEL_LOGFILENAME");
			if (fn && strlen(fn)) {
				freopen(fn, "a", stdout);
				freopen(fn, "a", stderr);
			}
			continue;
		}
		else if (strncmp(metadata[0], "@@idle", 6) == 0) {
			/* Ignored */
		}
		else {
			errprintf("Dropping message type %s, metacount=%d\n", metadata[0], metacount);
		}

		MEMUNDEFINE(logfn);
	}

	return 0;
}

